functional/generics-type-classes.md

```diff
@@ -0,0 +1,268 @@
+# Generics as Type Classes
+
+## Description
+
+Rust's type system is designed more like functional languages (like Haskell)
+rather than imperative languages (like Java and C++). As a result, Rust can turn
+many kinds of programming problems into "static typing" problems. This is one
+of the biggest wins of choosing a functional language, and is critical to many
+of Rust's compile time guarantees.
+
+A key part of this idea is the way generic types work. In C++ and Java, for
+example, generic types are a meta-programming construct for the compiler.
+`Vec<int>` and `Vec<char>` in C++ are just two different copies of the same
+boilerplate code for a `Vec` type (known as a `template`)  with two
```

Comment:
```suggestion
`vector<int>` and `vector<char>` in C++ are just two different copies of the same
boilerplate code for a `vector` type (known as a `template`)  with two
```
Should we use cpp names?

Reply:
Absolutely! Brain fade on my part :smile: